src/NHibernate/Util/AsyncReaderWriterLock.cs

```diff
@@ -191,13 +208,16 @@ private void ExitReadLock()
 
 		private void DisposeWaitingSemaphore()
 		{
-			if (_waitingDisposalSemaphore == null)
+			_waitingDisposalSemaphore?.Dispose();
+			_waitingDisposalSemaphore = null;
+		}
+
+		private void AssertNotDisposed()
+		{
+			if (_disposed)
 			{
-				return;
+				throw new InvalidOperationException("The instance is disposed.");
```

Comment:
Why not `throw new ObjectDisposedException(nameof(AsyncReaderWriterLock));`?

Reply:
My bad, corrected.